tests/unit/state_manager/test_state_manager.py

```diff
@@ -222,6 +222,94 @@ def test_set_state(self):
                 ran = sm.has_step_run(part, step)
                 assert ran == (part == p2 and step == Step.BUILD)
 
+    def test_should_step_run_trivial(self):
+        info = ProjectInfo()
+        p1 = Part("p1", {})
+
+        sm = StateManager(project_info=info, part_list=[p1])
+
+        for step in list(Step):
+            assert sm.should_step_run(p1, step) is True
+
+    def test_should_step_run_step_already_ran(self):
+        info = ProjectInfo()
+        p1 = Part("p1", {})
+
+        # p1 pull already ran
+        s1 = states.StageState()
+        s1.write(Path("parts/p1/state/pull"))
+
+        sm = StateManager(project_info=info, part_list=[p1])
+
+        for step in list(Step):
+            assert sm.should_step_run(p1, step) == (step != Step.PULL)
+
+    def test_should_step_run_outdated(self):
+        info = ProjectInfo()
+        p1 = Part("p1", {})
+
+        # p1 build already ran
+        s1 = states.BuildState()
+        s1.write(Path("parts/p1/state/build"))
+
+        # but p1 pull ran more recently
+        time.sleep(0.1)
```

Comment:
```suggestion
        time.sleep(0.001)
```
or something that meets the required granularity :D

Reply:
Probably any small interval in the millisecond range would work. Will decrease to the suggested value.

Reply:
Interestingly enough, something is not right when you use intervals < 5ms, as you can see in the code vs the actual mtime timestamps from the files:
```
    s4.write(Path("parts/bar/state/prime"))                                                                                                                                              
    time.sleep(0.001)                                                                                                                                                                    
    s3.write(Path("parts/bar/state/pull"))                                                                                                                                               
    time.sleep(0.001)                                                                                                                                                                    
    s1.write(Path("parts/foo/state/pull"))                                                                                                                                               
    time.sleep(0.001)                                                                                                                                                                    
    s2.write(Path("parts/foo/state/build"))             

 foo Step.PULL   1618444454171108403                                                        
 foo Step.BUILD  1618444454171108403                                                                                                                                                      
 bar Step.PULL   1618444454167108375    
 bar Step.PRIME  1618444454167108375
```
Either Python's `time.sleep()` call is not very accurate, or actual file writing is being delayed somewhere and happening in bursts. After a short investigation I realized that I'm running it on Ubuntu Server with HZ=250, which gives me 4ms resolution (matching the 167 to 171ms interval in the timestamps). So I'll use the more conservative value of 20ms, which should be more than enough for systems running with HZ=100.